ssmpu: refine ssmpu with suggestions from TG

Ddnirvana opened the pull request:
The PR refines MPU spec with Nick and Herinch's comments.

Specifically, it includes the following updates:

* ssmpu: refer PMP/ePMP's description about address matching
* support upto 64 MPU entries
* refine description about contiguous bits 
* refine permission encodings using rule-based terms
* update figs and refine descriptions
* put supports for hypervisor in a new section

All the updates are also applied on gdoc too.

Ddnirvana committed:
ssmpu: refine expression

Signed-off-by: Dong Du <Dd_nirvana@sjtu.edu.cn>

Ddnirvana committed:
ssmpu: refer PMP/ePMP's description about address matching

Signed-off-by: Dong Du <Dd_nirvana@sjtu.edu.cn>

Ddnirvana committed:
ssmpu: support upto 64 MPU entries

Signed-off-by: Dong Du <Dd_nirvana@sjtu.edu.cn>

Ddnirvana committed:
ssmpu: put supports for hypervisor in a new section

Signed-off-by: Dong Du <Dd_nirvana@sjtu.edu.cn>

Ddnirvana committed:
ssmpu: refine description about contiguous bits

Signed-off-by: Dong Du <Dd_nirvana@sjtu.edu.cn>

Ddnirvana committed:
ssmpu: refine permission encodings using rule-based terms

Signed-off-by: Dong Du <Dd_nirvana@sjtu.edu.cn>

Ddnirvana committed:
ssmpu: refine explanation of SUM bit

Signed-off-by: Dong Du <Dd_nirvana@sjtu.edu.cn>

Ddnirvana committed:
ssmpu: update figs and refine descriptions

Signed-off-by: Dong Du <Dd_nirvana@sjtu.edu.cn>

Ddnirvana committed:
ssmpu: refine minor typos

Signed-off-by: Dong Du <Dd_nirvana@sjtu.edu.cn>

Ddnirvana committed:
merge with main

Signed-off-by: Dong Du <Dd_nirvana@sjtu.edu.cn>

mickflemm commented on Ssmpu/Memory_Protection_Unit.adoc:
```diff
@@ -5,14 +5,16 @@ An optional RISC-V Memory Protection Unit (MPU) provides per-hart supervisor-mod
 Like PMP, the granularity of MPU access control settings are platform-specific and within a platform may vary by physical memory region, but the standard MPU encoding should support regions as small as four bytes. 
 
-MPU checks could be applied to all accesses for both U mode and S mode, depending on the values in the configuration registers. MPU registers can always be modified by M-mode and S-mode software. MPU registers can grant permissions to U-mode, which has none by default, and revoke permissions from S-mode, which has full permissions by default.
+MPU checks could be applied to all accesses for both U mode and S mode, depending on the values in the configuration registers.
```
could be applied or will be applied ?

Ddnirvana replied:
Yes. It would be better to use "will be applied".

mickflemm commented on Ssmpu/Memory_Protection_Unit.adoc:
```diff
@@ -5,14 +5,16 @@ An optional RISC-V Memory Protection Unit (MPU) provides per-hart supervisor-mod
-MPU checks could be applied to all accesses for both U mode and S mode, depending on the values in the configuration registers. MPU registers can always be modified by M-mode and S-mode software. MPU registers can grant permissions to U-mode, which has none by default, and revoke permissions from S-mode, which has full permissions by default.
+MPU checks could be applied to all accesses for both U mode and S mode, depending on the values in the configuration registers.
+M-mode accesses are not affected.
+MPU registers can always be modified by M-mode and S-mode software. MPU registers can grant permissions to U-mode, which has none by default, and revoke permissions from S-mode, which has full permissions by default.
```
let's make it clear that "full permissions" in this case is whatever M-mode allowed through PMP/ePMP.

Ddnirvana replied:
How about replace "which has full permissions by default." to "which has all permissions allowed through PMP/ePMP."?
That means by default the S-mode has all permission indicated by PMP/ePMP, but MPU can further restrict the permission.

mickflemm commented on Ssmpu/Memory_Protection_Unit.adoc:
```diff
@@ -29,7 +29,8 @@ image::RV64_MPU_configuration_CSR_layout.png[title="RV64 MPU configuration CSR l
 The MPU address registers are CSRs named mpuaddr0-mpuaddr63.
 Each MPU address register encodes bits 33-2 of 34-bit physical address for RV32, as shown in Figure 4.
-For RV64, each MPU address encodes bits 55–2 of a 56-bit physical address, as shown in Figure 5. Not all physical address bits may be implemented (but implemented RW bits must be contiguous), and so the mpuaddr registers are WARL.
+Fewer address bits may be implemented for systems with a smaller physical address space.
```
Missed the previous sentence about RV64 "each MPU address encodes bits 55 - 2..."

Ddnirvana replied:
Thanks for pointing it out! Will fix it.

allenjbaum replied:
For what it's worth, fewer address bits may be implemented even if the physical address space isn't smaller, e.g. the if the minimum grain size is >2 so low order bits are read-only zeroes AND TOR mode is not legal. PMP has some very complicated rules about about what happens if you switch an entry from TOR to NAPOT and then back to TOR. If the grain size forces the low order bits to zero, and then back, those low order bits are retained. But if TOR entries are not implemented (which is a legal configuration since the A field is WARL) then those bits don't have to be implemented at all.

Ddnirvana replied:
@allenjbaum Thanks for pointing it out.
Will change to "Fewer address bits may be implemented for specific reasons, e.g., systems have a smaller physical address space."

mickflemm commented on Ssmpu/Memory_Protection_Unit.adoc:
```diff
@@ -37,9 +37,12 @@ image::RV32_MPU_address_register_format.png[title="MPU address register format,
 
 The layout of MPU configuration registers is the same as PMP configuration registers, as is shown in Figure 6.
-The X/W/R field controls instruction execution, write and read permissions.
+The X/W/R bitsd controls instruction execution, write and read permissions.
```
bitsd -> bits

Ddnirvana replied:
Thanks for pointing it out, will fix it.

mickflemm commented on Ssmpu/Memory_Protection_Unit.adoc:
```diff
@@ -37,9 +37,12 @@ image::RV32_MPU_address_register_format.png[title="MPU address register format,
+The X/W/R bitsd controls instruction execution, write and read permissions.
+
+The S bit marks a rule as *S-mode-only* when set and *U-mode-only* when unset.
+The reserved encoding of ``mpucfg.RW=01``, and the encoding ``mpucfg.SRWX=1111``, now encode a Shared-Region.
```
those weren't reserved before, mpucfg registers are defined fom scratch on this spec

mickflemm replied:
we also need to define what an entry/rule is (or maybe just say it's similar to ePMP)

Ddnirvana replied:
I will remove the "reserved" in the next commit.
I prefer to say the definition of entry/rule is similar to ePMP.

allenjbaum replied:
Pedantic: many orderings of RWX and XWR are used; could you settle for just one?

Ddnirvana replied:
@allenjbaum Thanks Allen. Will use RWX only in the whole spec.

mickflemm commented on Ssmpu/Memory_Protection_Unit.adoc:
```diff
@@ -57,13 +60,29 @@ The A field in an MPU entry’s configuration register encodes the address-match
-S-mode can always modify the value in S bit.
+MPU has three kinds of rules: *U-mode-only*, *S-mode-only*, and *Shared-Region* rules.
+The S bit marks a rule as *S-mode-only* when set and *U-mode-only* when unset.
+The encoding ``mpucfg.SRWX=1000`` is reserved and the encoding ``mpucfg.RW=01`` and the encoding ``mpucfg.SRWX=1111`` encodes a Shared-Region.
```
no need to mention RW=01 or SRWX=111 here

mickflemm replied:
let's clean this up a bit, let's say something like "mpucfg.RW=01  encodes a shared region and mpucfg.SRWX=1000 is reserved for future standard use". mpucfg.SRWX=1111 is mentioned later on, and since it's a special case let's leave it there.

mickflemm commented on Ssmpu/Memory_Protection_Unit.adoc:
```diff
@@ -57,13 +60,29 @@ The A field in an MPU entry’s configuration register encodes the address-match
++
+. A _U-mode-only_ rule is *enforced* on User modes and *denied*/*enforced* on Supervisor mode depending on the value of ``satp.SUM`` bit:
++
+* If ``satp.SUM`` is set, a U-mode-only rule is enforced yet non-executable on Supervisor mode in order to ensure SMEP.
```
we haven't defined a non-executable rule, let's just say without code execution privileges

Ddnirvana replied:
Reasonable. Thx.

allenjbaum replied:
The SUM bit  is in the SSTATUS CSR, not SATP !

Ddnirvana replied:
> The SUM bit is in the SSTATUS CSR, not SATP !

@allenjbaum Hi Allen, the shown code seems outdated. It has been updated to `sstatus.SUM` during the PR created.

mickflemm commented on Ssmpu/Memory_Protection_Unit.adoc:
```diff
@@ -57,13 +60,29 @@ The A field in an MPU entry’s configuration register encodes the address-match
++
+. A _Shared-Region_ rule is enforced on both Supervisor and User modes, with restrictions depending on the ``mpucfg.S`` and ``mpucfg.X`` bits:
++
+* If ``mpucfg.S`` is not set the region can be used for sharing data between S-mode and U-mode so is not executable. S-mode has RW access to that region and U-mode has read access if ``mpucfg.X`` is not set, or RW access if ``mpucfg.X`` is set.
```
so is not -> but can't be

mickflemm replied:
also let's say read-only access if mpucfg.X is not set

mickflemm commented on Ssmpu/Memory_Protection_Unit.adoc:
```diff
@@ -102,12 +121,12 @@ MPU does not permit supervisor mode to execute instructions from U-mode regions,
 The PMP checks only take effect after the memory access passes the MPU permission checks. An M-mode access will not be checked by MPU property.
 
-Like PMP entries, MPU entries are also statically prioritized. The lowest-numbered MPU entry that matches any byte of an access determines whether that access succeeds or fails. The matching MPU entry must match all bytes of an access, or the access fails, irrespective of the S, R, W, and X bits.
+Like PMP entries, MPU entries are also statically prioritized. The lowest-numbered MPU entry that matches any byte of an access determines whether that access is allowed or fails. The matching MPU entry must match all bytes of an access, or the access fails, irrespective of the S, R, W, and X bits.
```
what do you mean by "any byte of an access" ? entries are matched using addresses, did you mean any byte of an address ?

Ddnirvana replied:
It is any bytes represented by an address and the length (indicated by the load/store instructions).

Ddnirvana committed:
ssmpu: refine ssmpu

Signed-off-by: Dong Du <Dd_nirvana@sjtu.edu.cn>

Ddnirvana committed:
ssmpu: refine with Allen's comments

Signed-off-by: Dong Du <Dd_nirvana@sjtu.edu.cn>

Ddnirvana committed:
ssmpu: refine with Allen's comments

Signed-off-by: Dong Du <Dd_nirvana@sjtu.edu.cn>

Ddnirvana committed:
ssmpu: refine MPU and Paging section

As suggested by Bill, we clarify the conditions to use MPU/MMU in
Section 2.6.

Signed-off-by: Dong Du <Dd_nirvana@sjtu.edu.cn>

Ddnirvana committed:
ssmpu: clarify the checking order between MPU and PMP/PMA

Signed-off-by: Dong Du <Dd_nirvana@sjtu.edu.cn>
